internal/pkg/service/appsproxy/http/router.go

```diff
@@ -63,19 +64,40 @@ func NewRouter(d dependencies.ServiceScope, exceptionIDPrefix string) (*Router,
 	}
 
 	router := &Router{
-		logger:            d.Logger(),
-		telemetry:         d.Telemetry(),
-		config:            d.Config(),
-		clock:             d.Clock(),
-		loader:            d.Loader(),
-		appHandlers:       make(map[string]http.Handler),
+		logger:    d.Logger(),
+		telemetry: d.Telemetry(),
+		config:    d.Config(),
+		clock:     d.Clock(),
+		loader:    d.Loader(),
+		appHandlers: appconfig.NewSafeMap[string, appHandler](func() *appHandler {
+			return &appHandler{
+				updateLock: &sync.RWMutex{},
+			}
+		}),
 		selectionTemplate: tmpl,
 		exceptionIDPrefix: exceptionIDPrefix,
 	}
 
 	return router, nil
 }
 
+type appHandler struct {
+	httpHandler http.Handler
+	updateLock  *sync.RWMutex
+}
+
+func (h *appHandler) getHTTPHandler() http.Handler {
+	h.updateLock.RLock()
+	defer h.updateLock.RUnlock()
+	return h.httpHandler
+}
+
+func (h *appHandler) setHTTPHandler(httpHandler http.Handler) {
+	h.updateLock.Lock()
+	defer h.updateLock.Unlock()
+	h.httpHandler = httpHandler
+}
+
 func (r *Router) CreateHandler() http.Handler {
 	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
 		appIDString, ok := ctxattr.Attributes(req.Context()).Value(attrAppID)
```

Comment:
Suggestion:
- From the code, I don't know where the `appID` is set/defined.
- Telemetry attributes are not intended for values passing.
- You can use something like:
```
ctx = context.WithValue(ctx, appIDCtxKey, appID)

....

appID, ok := ctx.Value(appIDCtxKey).(string)
```